File: Spiral_Matrix_II.java
/*59. Spiral Matrix II Add to List
Description  Submission  Solutions
Total Accepted: 74340
Total Submissions: 193550
Difficulty: Medium
Contributors: Admin
Given an integer n, generate a square matrix filled with elements from 1 to n2 in spiral order.

For example,
Given n = 3,

You should return the following matrix:
[
 [ 1, 2, 3 ],
 [ 8, 9, 4 ],
 [ 7, 6, 5 ]
]
Hide Tags Array
Hide Similar Problems (M) Spiral Matrix
 */






/*
(3 ms/ 21 test)
four index
int right = 0;
int down = 0;
int left = spiral[0].length-1;
int up = spiral.length-1;
use num to help count the number should be in spiral
int num = 1;
according to the order, right down left up
 */

////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public int[][] generateMatrix(int n) {
        int[][] spiral = new int[n][n];
        if(n==0) return spiral;

        int num = 1;
        int right = 0;
        int down = 0;
        int left = spiral[0].length-1;
        int up = spiral.length-1;

        while(left>=right && up>=down){
            for(int i = right; i <= left; ++i){
                spiral[down][i] = num;
                num++;
            }
            down++;

            for(int i = down; i <= up; ++i){
                spiral[i][left] = num;
                num++;
            }
            left--;

            if(down<=up){
                for(int i = left; i >= right; --i){
                    spiral[up][i] = num;
                    num++;
                }
                up--;
            }

            if(right<=left){
                for(int i = up; i >= down; --i){
                    spiral[i][right] = num;
                    num++;
                }
                right++;
            }
        }
        return spiral;
    }
}

File: Spiral_Matrix_II.py
'''
Given a positive integer n, generate an n x n matrix filled with elements from 1 to n2 in spiral order.



Example 1:


Input: n = 3
Output: [[1,2,3],[8,9,4],[7,6,5]]
Example 2:

Input: n = 1
Output: [[1]]


Constraints:

1 <= n <= 20
'''

'''
Solution 1:
build matrix element one by one
use dir to help define the direction of each element
0 right, 1 down, 2 left, 3 up

Time Complexity: O(n^2)
Space Complexity: O(n^2)
'''
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        matrix = [[0 for i in range(n)] for i in range(n)]

        # 0 right, 1 down, 2 left, 3 up
        d = 0

        i = 0
        j = 0
        for val in range(1, n*n+1):
            matrix[i][j] = val
            if d == 0:
                if j+1 == n or matrix[i][j+1] != 0:
                    # change direction to down
                    d = 1
                    i += 1
                else:
                    j += 1
            elif d == 1:
                if i+1 == n or matrix[i+1][j] != 0:
                    # change direction to left
                    d = 2
                    j -= 1
                else:
                    i += 1
            elif d == 2:
                if j-1 < 0 or matrix[i][j-1] != 0:
                    # change direction to up
                    d = 3
                    i -= 1
                else:
                    j -= 1
            elif d == 3:
                if i-1 < 0 or matrix[i-1][j] != 0:
                    # change direction to right
                    d = 0
                    j += 1
                else:
                    i -= 1

        return matrix
